<a href="https://colab.research.google.com/github/sharmawiki/genai/blob/main/openai_with_langchain_and_pinecone.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://github.com/sharmawiki/genai/blob/main/openai_with_langchain_and_pinecone.ipynb) [![Open In Cogniverse.net](https://cogniverse.net/wp-content/uploads/cogniverse_logo_gradient_blue_80x20.jpg)](https://cogniverse.net/guide-to-openai-embeddings-integration-with-langchain-pinecone/)

#### OpenAI with Langchain and Pinecone

#Install Necessary Dependencies

In [ ]:
!pip install -qU \
    openai==1.30.1 \
    pinecone-client==4.1.0 \
    langchain==0.1.20 \
    langchain-openai==0.1.6 \
    langchain-community==0.0.38 \
    tiktoken==0.7.0 \
    datasets==2.19.1

## Prep The Data For Use
### Load Dataset
Next, we require a dataset for our exercise. We’ll use a publicly available dataset from Hugging Face, as shown below:

In [ ]:
from datasets import load_dataset

data = load_dataset('squad', split='train')
data

Dataset({
    features: ['id', 'title', 'context', 'question', 'answers'],
    num_rows: 87599
})

### Transform Data To Pandas DataFrame
Once the data is loaded, we’ll convert it into a Pandas DataFrame

In [ ]:
data = data.to_pandas()
data.head()

id                     title  \
0  5733be284776f41900661182  University_of_Notre_Dame   
1  5733be284776f4190066117f  University_of_Notre_Dame   
2  5733be284776f41900661180  University_of_Notre_Dame   
3  5733be284776f41900661181  University_of_Notre_Dame   
4  5733be284776f4190066117e  University_of_Notre_Dame   

                                             context  \
0  Architecturally, the school has a Catholic cha...   
1  Architecturally, the school has a Catholic cha...   
2  Architecturally, the school has a Catholic cha...   
3  Architecturally, the school has a Catholic cha...   
4  Architecturally, the school has a Catholic cha...   

                                            question  \
0  To whom did the Virgin Mary allegedly appear i...   
1  What is in front of the Notre Dame Main Building?   
2  The Basilica of the Sacred heart at Notre Dame...   
3                  What is the Grotto at Notre Dame?   
4  What sits on top of the Main Building at Notre...   

                                             answers  
0  {'text': ['Saint Bernadette Soubirous'], 'answ...  
1  {'text': ['a copper statue of Christ'], 'answe...  
2  {'text': ['the Main Building'], 'answer_start'...  
3  {'text': ['a Marian place of prayer and reflec...  
4  {'text': ['a golden statue of the Virgin Mary'...

### Remove Duplicates From Dataset
Upon inspecting the data table, you may notice duplicates in the ‘title’ and ‘context’ columns. We can remove these duplicates using the following code:

In [ ]:
data.drop_duplicates(subset='context', keep='first', inplace=True)
data.head()

id                     title  \
0   5733be284776f41900661182  University_of_Notre_Dame   
5   5733bf84d058e614000b61be  University_of_Notre_Dame   
10  5733bed24776f41900661188  University_of_Notre_Dame   
15  5733a6424776f41900660f51  University_of_Notre_Dame   
20  5733a70c4776f41900660f64  University_of_Notre_Dame   

                                              context  \
0   Architecturally, the school has a Catholic cha...   
5   As at most other universities, Notre Dame's st...   
10  The university is the major seat of the Congre...   
15  The College of Engineering was established in ...   
20  All of Notre Dame's undergraduate students are...   

                                             question  \
0   To whom did the Virgin Mary allegedly appear i...   
5   When did the Scholastic Magazine of Notre dame...   
10  Where is the headquarters of the Congregation ...   
15  How many BS level degrees are offered in the C...   
20  What entity provides help with the management ...   

                                              answers  
0   {'text': ['Saint Bernadette Soubirous'], 'answ...  
5   {'text': ['September 1876'], 'answer_start': [...  
10          {'text': ['Rome'], 'answer_start': [119]}  
15         {'text': ['eight'], 'answer_start': [487]}  
20  {'text': ['Learning Resource Center'], 'answer...

## Prepare Environment (OpenAI, Pinecone & Langchain)
Now that our data is prepared, we can utilize OpenAI embeddings with LangChain and Pinecone. At this stage, you will need:

* An OpenAI API Key (charges may apply as OpenAI services are now paid).
* A Pinecone API Key (used to interact with the vector database).


### OpenAI Readiness

In [ ]:
import os
from getpass import getpass
from langchain_openai import OpenAIEmbeddings

# get API key from top-right dropdown on OpenAI website
OPENAI_API_KEY = os.getenv("OPENAI_API_KEY") or getpass("Enter your OpenAI API key: ")
model_name = 'text-embedding-ada-002'

embed = OpenAIEmbeddings(
    model=model_name,
    openai_api_key=OPENAI_API_KEY
)

Enter your OpenAI API key: ··········


### Pinecone Readiness
#### Initialize PINECONE Client
Next we will use PINECONE API key to initialize our Pinecone client using the code below

In [ ]:
from pinecone import Pinecone

# initialize connection to pinecone (get API key at app.pinecone.io)
api_key = os.getenv("PINECONE_API_KEY") or getpass("Enter your Pinecone API key: ")

# configure client
pc = Pinecone(api_key=api_key)

Enter your Pinecone API key: ··········


#### Define PINECONE Index Specs


In [ ]:
from pinecone import ServerlessSpec

spec = ServerlessSpec(
    cloud="aws", region="us-east-1"
)

## Indexing Our Dataset In Pinecone
### Create Index
We need to create a Pinecone index based on our requirements. Since we’re using the ‘text-embedding-ada-002’ model,

* dimension, should be equal to to dimensionality of Ada-002 (`1536`), and
* metric, can be either `cosine` or `dotproduct`

In [ ]:
import time

index_name = "langchain-retrieval-agent"
existing_indexes = [
    index_info["name"] for index_info in pc.list_indexes()
]

# check if index already exists (it shouldn't if this is first time)
if index_name not in existing_indexes:
    # if does not exist, create index
    pc.create_index(
        index_name,
        dimension=1536,  # dimensionality of ada 002
        metric='dotproduct',
        spec=spec
    )
    # wait for index to be initialized
    while not pc.describe_index(index_name).status['ready']:
        time.sleep(1)

# connect to index
index = pc.Index(index_name)
time.sleep(1)
# view index stats
index.describe_index_stats()

{'dimension': 1536,
 'index_fullness': 0.0,
 'namespaces': {},
 'total_vector_count': 0}

### Indexing With Pinecone
Now, we’ll use the Pinecone index to start indexing our dataset:

In [ ]:
from tqdm.auto import tqdm

batch_size = 100

texts = []
metadatas = []

for i in tqdm(range(0, len(data), batch_size)):
    # get end of batch
    i_end = min(len(data), i+batch_size)
    batch = data.iloc[i:i_end]
    # first get metadata fields for this record
    metadatas = [{
        'title': record['title'],
        'text': record['context']
    } for j, record in batch.iterrows()]
    # get the list of contexts / documents
    documents = batch['context']
    # create document embeddings
    embeds = embed.embed_documents(documents)
    # get IDs
    ids = batch['id']
    # add everything to pinecone
    index.upsert(vectors=zip(ids, embeds, metadatas))

  0%|          | 0/189 [00:00<?, ?it/s]

#### Confirm Indexing
Wait until the indexing finishes. Once done, confirm by executing the following code in the notebook:

In [ ]:
index.describe_index_stats()

{'dimension': 1536,
 'index_fullness': 0.0,
 'namespaces': {'': {'vector_count': 18891}},
 'total_vector_count': 18891}

If you have used the same dataset, then the total_vector_count should be 18891.



## Vectorstore With Langchain
Now, with our vector database ready, let’s initialize our Vector store with Langchain:



In [ ]:
from langchain.vectorstores import Pinecone

text_field = "text"  # the metadata field that contains our text

# initialize the vector store object
vectorstore = Pinecone(
    index, embed.embed_query, text_field
)

## Let’s Search (Important Checkpoint)
After all the preparations, it’s time to execute our first search:

In [ ]:
query = "when was the college of engineering in the University of Notre Dame established?"

vectorstore.similarity_search(
    query,  # our search query
    k=3  # return 3 most relevant docs
)

[Document(page_content="In 1919 Father James Burns became president of Notre Dame, and in three years he produced an academic revolution that brought the school up to national standards by adopting the elective system and moving away from the university's traditional scholastic and classical emphasis. By contrast, the Jesuit colleges, bastions of academic conservatism, were reluctant to move to a system of electives. Their graduates were shut out of Harvard Law School for that reason. Notre Dame continued to grow over the years, adding more colleges, programs, and sports teams. By 1921, with the addition of the College of Commerce, Notre Dame had grown from a small college to a university with five colleges and a professional law school. The university continued to expand and add new residence halls and buildings with each subsequent president.", metadata={'title': 'University_of_Notre_Dame'}),
 Document(page_content='The College of Engineering was established in 1920, however, early c

Lets continue this success in our next notebook.